c_src/xav/xav_video_converter.c

```diff
@@ -0,0 +1,136 @@
+#include "xav_video_converter.h"
+
+ErlNifResourceType * xav_video_converter_resource_type;
+
+ERL_NIF_TERM new(ErlNifEnv * env, int argc, const ERL_NIF_TERM argv[]) {
+  if (argc != 4) {
+    return xav_nif_error(env, "invalid_arg_count");
+  }
+
+  ERL_NIF_TERM ret;
+
+  int in_width;
+  if (!enif_get_int(env, argv[0], &in_width)) {
+    return xav_nif_raise(env, "failed_to_get_int");
+  }
+
+  int in_height;
+  if (!enif_get_int(env, argv[1], &in_height)) {
+    return xav_nif_raise(env, "failed_to_get_int");
+  }
+
+  unsigned int in_format_len;
+  if (!enif_get_atom_length(env, argv[2], &in_format_len, ERL_NIF_LATIN1)) {
+    return xav_nif_raise(env, "failed_to_get_atom_length");
+  }
+
+  char *in_format = (char *)XAV_ALLOC((in_format_len + 1) * sizeof(char *));
+  if (enif_get_atom(env, argv[2], in_format, in_format_len + 1, ERL_NIF_LATIN1) == 0) {
+    return xav_nif_raise(env, "failed_to_get_atom");
+  }
+
+  unsigned int out_format_len;
+  if (!enif_get_atom_length(env, argv[3], &out_format_len, ERL_NIF_LATIN1)) {
+    return xav_nif_raise(env, "failed_to_get_atom_length");
+  }
+
+  char *out_format = (char *)XAV_ALLOC((out_format_len * 1) * sizeof(char *));
+  if (enif_get_atom(env, argv[3], out_format, out_format_len + 1, ERL_NIF_LATIN1) == 0) {
+    return xav_nif_raise(env, "failed_to_get_atom");
+  }
+
+  enum AVPixelFormat in_pix_fmt = av_get_pix_fmt(in_format);
+  enum AVPixelFormat out_pix_fmt = av_get_pix_fmt(out_format);
+
+  if (in_pix_fmt == AV_PIX_FMT_NONE || out_pix_fmt == AV_PIX_FMT_NONE) {
+    ret = xav_nif_error(env, "unknown_format");
+    goto fail;
+  }
+
+  struct XavVideoConverter *xav_video_converter = enif_alloc_resource(xav_video_converter_resource_type, 
+                                                                      sizeof(xav_video_converter));
+  xav_video_converter->vc = video_converter_alloc();
+  if (video_converter_init(xav_video_converter->vc, in_width, in_height, in_pix_fmt, out_pix_fmt) < 0) {
+    return xav_nif_raise(env, "failed_to_init_converter");
+  }
```

Comment:
Do this on converting the first video frame. This way we can get rid of `in_*` arguments and make API cleaner